### PR TITLE
fix: password rotation policy

### PR DESCRIPTION
**Pull request by tenthirtyam**
### Summary

- Updates `Invoke-PasswordRotationManager` documentation to note that the `sddcRootPass` parameter is required.
- Updates `Invoke-PasswordRotationManager` to ensure that the workload domain passed to `-workloadDomain` parameter is present in the inventory. If not, stop with an error message.
- Updates `Invoke-PasswordRotationManager` to only run policy checks for SDDC Manager and vCenter Single Sign-on if `-allDomains` is passed or if `-workloadDomain` is the correct type.
- Updates `Invoke-PasswordRotationManager` to check if Aria Suite Lifecycle is integrated before proceeding with the Aria checks. Only returns results for those products that are in the SDDC Manager inventory.
- Updates `Invoke-PasswordRotationManager` to only check the Aria Suite if `-allDomaina` is passed or the `-workloadDomain` is the correct type.
- If a workload does not return data for an object, such as no SDDC Manager managed NSX Edges in a workload domain, the `Publish-PasswordRotationPolicy` cmdlet will present a message on the HTML output.
- Updates `Save-ClarityReportNavigationForRotation` to generate the navigation based on the Aria Suite products in the SDDC Manager inventory.
- Updates module version to 1.5.0.1003.

### Type

- [x] Bugfix
- [ ] Enhancement or Feature
- [ ] Code Style or Formatting
- [ ] Documentation
- [ ] Refactoring
- [ ] Chore
- [ ] Other
        Please describe:

### Breaking Changes?

- [ ] Yes, there are breaking changes.
- [x] No, there are no breaking changes.

### Test and Documentation

- [x] Tests have been completed.
- [x] Documentation has been added or updated.

```powershell
PS F:\>  Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! -reportPath "F:\Reporting" -darkMode -allDomains -sddcRootPass VMw@re1!

[10-09-2023_19:44:48] INFO Starting the Process of Generating Password Rotation Manager Report for VMware Cloud Foundation instance (sfo-vcf01.sfo.rainpole.io).
 [10-09-2023_19:44:48] INFO Setting up the log file to path F:\Reporting\logs\Invoke-PasswordRotationManager-10-09-2023_07_44_48.log.
 [10-09-2023_19:44:48] INFO Setting up report folder and report F:\Reporting\PasswordRotationManager\10-09-2023_07_44_48-passwordRotationManager-sfo-vcf01.htm.
 [10-09-2023_19:44:49] INFO Collecting SDDC Manager password rotation policy for VMware Cloud Foundation instance (sfo-vcf01.sfo.rainpole.io).
 [10-09-2023_19:44:50] INFO Collecting vCenter Single Sign-On password rotation policy for VMware Cloud Foundation instance (sfo-vcf01.sfo.rainpole.io).
 [10-09-2023_19:44:52] INFO Collecting vCenter Server password rotation policy for VMware Cloud Foundation instance (sfo-vcf01.sfo.rainpole.io).
 [10-09-2023_19:44:53] INFO Collecting NSX Manager password rotation policy for VMware Cloud Foundation instance (sfo-vcf01.sfo.rainpole.io).
 [10-09-2023_19:44:54] INFO Collecting NSX Edge password rotation policy for VMware Cloud Foundation instance (sfo-vcf01.sfo.rainpole.io).
 [10-09-2023_19:44:56] INFO Collecting Aria Suite Lifecycle password rotation policy for VMware Cloud Foundation instance (sfo-vcf01.sfo.rainpole.io).
 [10-09-2023_19:44:57] INFO Collecting Aria Operations for Logs password rotation policy for VMware Cloud Foundation instance (sfo-vcf01.sfo.rainpole.io).
 [10-09-2023_19:44:58] INFO Skipping: Aria Operations is not present in the SDDC Manager inventory.
 [10-09-2023_19:44:58] INFO Skipping: Aria Automation is not present in the SDDC Manager inventory.
 [10-09-2023_19:44:58] INFO Skipping: Workspace ONE Access is not present in the SDDC Manager inventory.
 [10-09-2023_19:44:59] INFO Generating the Final Report and Saving to (F:\Reporting\PasswordRotationManager\10-09-2023_07_44_48-passwordRotationManager-sfo-vcf01.htm).
```

image

```powershell
PS F:\>  Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! -reportPath "F:\Reporting" -allDomains -sddcRootPass VMw@re1! -json

 [10-09-2023_19:47:05] INFO Starting the process of generating the password rotation JSON for VMware Cloud Foundation instance (sfo-vcf01.sfo.rainpole.io).
 [10-09-2023_19:47:06] INFO Collecting SDDC Manager password rotation policy for VMware Cloud Foundation instance (sfo-vcf01.sfo.rainpole.io).
 [10-09-2023_19:47:07] INFO Collecting vCenter Single Sign-On password rotation policy for VMware Cloud Foundation instance (sfo-vcf01.sfo.rainpole.io).
 [10-09-2023_19:47:09] INFO Collecting vCenter Server password rotation policy for VMware Cloud Foundation instance (sfo-vcf01.sfo.rainpole.io).
 [10-09-2023_19:47:10] INFO Collecting NSX Manager password rotation policy for VMware Cloud Foundation instance (sfo-vcf01.sfo.rainpole.io).
 [10-09-2023_19:47:12] INFO Collecting NSX Edge password rotation policy for VMware Cloud Foundation instance (sfo-vcf01.sfo.rainpole.io).
 [10-09-2023_19:47:13] INFO Collecting Aria Suite Lifecycle password rotation policy for VMware Cloud Foundation instance (sfo-vcf01.sfo.rainpole.io).
 [10-09-2023_19:47:15] INFO Collecting Aria Operations for Logs password rotation policy for VMware Cloud Foundation instance (sfo-vcf01.sfo.rainpole.io).
 [10-09-2023_19:47:16] INFO Skipping: Aria Operations is not present in the SDDC Manager inventory.
 [10-09-2023_19:47:16] INFO Skipping: Aria Automation is not present in the SDDC Manager inventory.
 [10-09-2023_19:47:16] INFO Skipping: Workspace ONE Access is not present in the SDDC Manager inventory.
 [10-09-2023_19:47:18] INFO Generating the Final JSON and Saving to (F:\Reporting\PasswordRotationManager\passwordRotationManager.json).
```

```powershell
PS F:\>  Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! -reportPath "F:\Reporting" -darkMode -workloadDomain sfo-m01 -sddcRootPass VMw@re1!

 [10-09-2023_19:47:52] INFO Starting the Process of Generating Password Rotation Manager Report for Workload Domain (sfo-m01).
 [10-09-2023_19:47:52] INFO Setting up the log file to path F:\Reporting\logs\Invoke-PasswordRotationManager-10-09-2023_07_47_51.log.
 [10-09-2023_19:47:52] INFO Setting up report folder and report F:\Reporting\PasswordRotationManager\10-09-2023_07_47_51-passwordRotationManager-sfo-m01.htm.
 [10-09-2023_19:47:52] INFO Collecting SDDC Manager password rotation policy for Workload Domain (sfo-m01).
 [10-09-2023_19:47:53] INFO Collecting vCenter Single Sign-On password rotation policy for Workload Domain (sfo-m01).
 [10-09-2023_19:47:53] INFO Collecting vCenter Server password rotation policy for Workload Domain (sfo-m01).
 [10-09-2023_19:47:54] INFO Collecting NSX Manager password rotation policy for Workload Domain (sfo-m01).
 [10-09-2023_19:47:54] INFO Collecting NSX Edge password rotation policy for Workload Domain (sfo-m01).
 [10-09-2023_19:47:54] INFO Collecting Aria Suite Lifecycle password rotation policy for Workload Domain (sfo-m01).
 [10-09-2023_19:47:55] INFO Collecting Aria Operations for Logs password rotation policy for Workload Domain (sfo-m01).
 [10-09-2023_19:47:55] INFO Skipping: Aria Operations is not present in the SDDC Manager inventory.
 [10-09-2023_19:47:55] INFO Skipping: Aria Automation is not present in the SDDC Manager inventory.
 [10-09-2023_19:47:55] INFO Skipping: Workspace ONE Access is not present in the SDDC Manager inventory.
 [10-09-2023_19:47:56] INFO Generating the Final Report and Saving to (F:\Reporting\PasswordRotationManager\10-09-2023_07_47_51-passwordRotationManager-sfo-m01.htm).
```

```powershell
PS F:\>  Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! -reportPath "F:\Reporting" -darkMode -workloadDomain sfo-m01 -sddcRootPass VMw@re1! -json

 [10-09-2023_19:48:48] INFO Starting the process of generating the password rotation JSON for Workload Domain (sfo-m01).
 [10-09-2023_19:48:49] INFO Collecting SDDC Manager password rotation policy for Workload Domain (sfo-m01).
 [10-09-2023_19:48:49] INFO Collecting vCenter Single Sign-On password rotation policy for Workload Domain (sfo-m01).
 [10-09-2023_19:48:50] INFO Collecting vCenter Server password rotation policy for Workload Domain (sfo-m01).
 [10-09-2023_19:48:50] INFO Collecting NSX Manager password rotation policy for Workload Domain (sfo-m01).
 [10-09-2023_19:48:51] INFO Collecting NSX Edge password rotation policy for Workload Domain (sfo-m01).
 [10-09-2023_19:48:51] INFO Collecting Aria Suite Lifecycle password rotation policy for Workload Domain (sfo-m01).
 [10-09-2023_19:48:51] INFO Collecting Aria Operations for Logs password rotation policy for Workload Domain (sfo-m01).
 [10-09-2023_19:48:52] INFO Skipping: Aria Operations is not present in the SDDC Manager inventory.
 [10-09-2023_19:48:52] INFO Skipping: Aria Automation is not present in the SDDC Manager inventory.
 [10-09-2023_19:48:52] INFO Skipping: Workspace ONE Access is not present in the SDDC Manager inventory.
 [10-09-2023_19:48:54] INFO Generating the Final JSON and Saving to (F:\Reporting\PasswordRotationManager\passwordRotationManager.json).
```

```powershell
PS F:\>  Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! -reportPath "F:\Reporting" -darkMode -workloadDomain sfo-w01 -sddcRootPass VMw@re1!

[10-09-2023_19:50:05] ERROR Workload Domain sfo-w01 not found. Please review the workload domain name and retry.
```

```powershell
PS F:\  Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! -reportPath "F:\Reporting" -darkMode -workloadDomain sfo-w02 -sddcRootPass VMw@re1!

 [10-09-2023_19:51:42] INFO Starting the Process of Generating Password Rotation Manager Report for Workload Domain (sfo-w02).
 [10-09-2023_19:51:42] INFO Setting up the log file to path F:\Reporting\logs\Invoke-PasswordRotationManager-10-09-2023_07_51_41.log.
 [10-09-2023_19:51:42] INFO Setting up report folder and report F:\Reporting\PasswordRotationManager\10-09-2023_07_51_41-passwordRotationManager-sfo-w02.htm.
 [10-09-2023_19:51:42] INFO Collecting vCenter Server password rotation policy for Workload Domain (sfo-w02).
 [10-09-2023_19:51:43] INFO Collecting NSX Manager password rotation policy for Workload Domain (sfo-w02).
 [10-09-2023_19:51:43] INFO Collecting NSX Edge password rotation policy for Workload Domain (sfo-w02).
 [10-09-2023_19:51:44] INFO Generating the Final Report and Saving to (F:\Reporting\PasswordRotationManager\10-09-2023_07_51_41-passwordRotationManager-sfo-w02.htm).
```

image

```powershell
PS F:\  Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! -reportPath "F:\Reporting" -darkMode -workloadDomain sfo-w02 -sddcRootPass VMw@re1! -json

 [10-09-2023_19:52:51] INFO Starting the process of generating the password rotation JSON for Workload Domain (sfo-w02).
 [10-09-2023_19:52:52] INFO Collecting vCenter Server password rotation policy for Workload Domain (sfo-w02).
 [10-09-2023_19:52:52] INFO Collecting NSX Manager password rotation policy for Workload Domain (sfo-w02).
 [10-09-2023_19:52:52] INFO Collecting NSX Edge password rotation policy for Workload Domain (sfo-w02).
 [10-09-2023_19:52:54] INFO Generating the Final JSON and Saving to (F:\Reporting\PasswordRotationManager\passwordRotationManager.json).
```

### Issue References

Related #122

### CODE REVIEW

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 1]
```diff
@@ -58,11 +64,11 @@ Function Invoke-PasswordRotationManager {
         The Invoke-PasswordRotationManager generates a Password Rotation Manager Report for a VMware Cloud Foundation instance.
 
         .EXAMPLE
-        Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! -reportPath "F:\Reporting" -darkMode -allDomains
+        Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! sddcRootPass VMw@re1! -reportPath "F:\Reporting" -darkMode -allDomains
```
Comment (bhumitra) [inline VMware.CloudFoundation.PasswordManagement.psm1:67]:
```suggestion
        Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! -sddcRootPass VMw@re1! -reportPath "F:\Reporting" -darkMode -allDomains
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 2]
```diff
@@ -58,11 +64,11 @@ Function Invoke-PasswordRotationManager {
...
         This example runs a password rotation report for all workload domains within an SDDC Manager instance.
 
         .EXAMPLE
-        Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1!  -reportPath "F:\Reporting" -darkMode -workloadDomain sfo-w01
+        Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! sddcRootPass VMw@re1! -reportPath "F:\Reporting" -darkMode -workloadDomain sfo-w01
```
Comment (bhumitra) [inline VMware.CloudFoundation.PasswordManagement.psm1:71]:
```suggestion
        Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! -sddcRootPass VMw@re1! -reportPath "F:\Reporting" -darkMode -workloadDomain sfo-w01
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 3]
```diff
@@ -4922,7 +5100,7 @@ Function Request-NsxtManagerPasswordComplexity {
                                             Write-Error "Unable to find Workload Domain typed (MANAGEMENT) in the inventory of SDDC Manager ($server): PRE_VALIDATION_FAILED"
                                         }
                                     }
-                                
+
```
Comment (bhumitra) [inline VMware.CloudFoundation.PasswordManagement.psm1:5103]:
```suggestion
```

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 4]
```diff
@@ -8775,7 +8953,7 @@ Function Request-LocalUserPasswordExpiration {
...
```
Comment (bhumitra) [inline VMware.CloudFoundation.PasswordManagement.psm1:8956]:
```suggestion
```

Remove empty line

File: VMware.CloudFoundation.PasswordManagement.psm1 [hunk 5]
```diff
@@ -9236,19 +9426,19 @@ Function Request-PasswordRotationPolicy {
...
+                    } catch {
+                    }
                     $passwordRotationExpiryDate = $passwordRotation.expiry.expiryDate
                     try {
                         $passwordRotationExpiryDate = [DateTime]::ParseExact($passwordRotation.expiry.expiryDate, 'yyyy-MM-ddTHH:mm:ss.fffZ', [System.Globalization.CultureInfo]::InvariantCulture)
-                    } catch {                                             
-                    } 
-    
+                    } catch {
+                    }
```
Comment (bhumitra) [inline VMware.CloudFoundation.PasswordManagement.psm1:9440]: Catch block is empty, add a log message.

File: docs/documentation/functions/Invoke-PasswordRotationManager.md [hunk 6]
```diff
@@ -27,15 +27,15 @@ The `Invoke-PasswordRotationManager` generates a Password Rotation Manager Repor
 ### Example 1
 
 ```powershell
-Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! -reportPath "F:\Reporting" -darkMode -allDomains        
+Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! sddcRootPass VMw@re1! -reportPath "F:\Reporting" -darkMode -allDomains
```
Comment (bhumitra) [inline docs/documentation/functions/Invoke-PasswordRotationManager.md:30]:
```suggestion
Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! -sddcRootPass VMw@re1! -reportPath "F:\Reporting" -darkMode -allDomains        
```

File: docs/documentation/functions/Invoke-PasswordRotationManager.md [hunk 7]
```diff
@@ -27,15 +27,15 @@ The `Invoke-PasswordRotationManager` generates a Password Rotation Manager Repor
...
+Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! sddcRootPass VMw@re1! -reportPath "F:\Reporting" -darkMode -allDomains        
 ```
 
 This example runs a password rotation report for all workload domains within an SDDC Manager instance.
 
 ### Example 2
 
 ```powershell
-Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1!  -reportPath "F:\Reporting" -darkMode -workloadDomain sfo-w01
+Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! sddcRootPass VMw@re1! -reportPath "F:\Reporting" -darkMode -workloadDomain sfo-w01
```
Comment (bhumitra) [inline docs/documentation/functions/Invoke-PasswordRotationManager.md:38]:
```suggestion
Invoke-PasswordRotationManager -sddcManagerFqdn sfo-vcf01.sfo.rainpole.io -sddcManagerUser admin@local -sddcManagerPass VMw@re1!VMw@re1! -sddcRootPass VMw@re1! -reportPath "F:\Reporting" -darkMode -workloadDomain sfo-w01
```

File: VMware.CloudFoundation.PasswordManagement.psm1 (hunk 3 above)
Comment (tenthirtyam) [reply VMware.CloudFoundation.PasswordManagement.psm1]: Resolved.

File: docs/documentation/functions/Invoke-PasswordRotationManager.md (hunk 6 above)
Comment (tenthirtyam) [reply docs/documentation/functions/Invoke-PasswordRotationManager.md]: Resolved.

File: docs/documentation/functions/Invoke-PasswordRotationManager.md (hunk 7 above)
Comment (tenthirtyam) [reply docs/documentation/functions/Invoke-PasswordRotationManager.md]: Resolved.

File: VMware.CloudFoundation.PasswordManagement.psm1 (hunk 4 above)
Comment (tenthirtyam) [reply VMware.CloudFoundation.PasswordManagement.psm1]: Resolved.

File: VMware.CloudFoundation.PasswordManagement.psm1 (hunk 1 above)
Comment (tenthirtyam) [reply VMware.CloudFoundation.PasswordManagement.psm1]: Resolved.

File: VMware.CloudFoundation.PasswordManagement.psm1 (hunk 5 above)
Comment (tenthirtyam) [reply VMware.CloudFoundation.PasswordManagement.psm1:9438]: Nice "catch"! I'll address this one shortly. 😀 

The other comments are resolved in the latest force push.

Comment (tenthirtyam) [reply VMware.CloudFoundation.PasswordManagement.psm1:9440]: Resolved in https://github.com/vmware/powershell-module-for-vmware-cloud-foundation-password-management/commit/c6c5af6dcb8e1f769c3432c4687b65dd40217af8.

Dropped in the use of `Debug-ExceptionWriter -object $_` in the two instances of an empty `] Catch [`.

**Comment by tenthirtyam** [comment]
Based on the image n the testing and provided below I see an issue that I'd like to address with the dynamic navigation.



The navigation for SDDC Manager and vCenter Single Sign-on should be excluded based on the targeted workload domain.

Standby for an update.

**Comment by tenthirtyam** [comment]
> Standby for an update.

Resolved in the latest push, but I had to revert the changes to the catch due to issues it was causing. We'll follow up with that item later.

image

**Review by bhumitra — APPROVED** [review]
LGTM

**Comment by tenthirtyam** [comment]
Attaching an example report.

[10-10-2023_06_59_05-passwordRotationManager-sfo-vcf01.htm.zip](https://github.com/vmware/powershell-module-for-vmware-cloud-foundation-password-management/files/12860436/10-10-2023_06_59_05-passwordRotationManager-sfo-vcf01.htm.zip)

**Pull request merged**